(#165) Removed all constructors except the last one

ruca905 opened the pull request:
Perhaps, I haven't understood the ticket, but i would be wonderful if somebody turn me in right direction

andreoss reviewed (changes requested):
@ruca905 it doesn't seem to be proper solution. Now there's a lot of copied code. 
I think you should file a new bug to clarify the intent of this one. 
@victornoel Probably, `TextMatcher` should have those constructors instead.

victornoel commented:
@andreoss @ruca905 I'm not extremely bothered by it, but I think this is because we should go farther than what we had before and implement for each matcher a dedicated description and mismatch function.

What is possible is that there is the need for an extra abstraction `Quoted implements Text` taking a `Text` in its constructor and quoting it (including escaping quotes in the text itself). This would be reused if there is the need to quote something (which is not always the case, for example you don't want to quote the regex.

@ruca905 you can start by at least writing some of the matcher's description and mismatch functions. If you can't do all of them in the allotted 30mn, then leave the required extra constructor and add a todo to finish the work :)

victornoel commented:
@ruca905 I see you pushed some changes, but that does not really is enough to merge this MR, I believe you could at least:
- rewrite some of the Matcher's to use the last constructor of MatcherOf
- be more detailed in the todo you modified 
- introduce the `QuotedText` class only if really needed, and I notice its implementation is incorrect since it does not quote anything
- Also `QuotedText` should contain a `@todo` to move it to cactoos and name it simply `Quoted` (as per the convention used in cactoos)

victornoel commented:
@ruca905 are you still working on this? :)

ruca905 commented:
Yes, I am very sorry for this delay, I will update the PR today, in a few hours.  But, to be honest, I don't fully understand the problem: what is quoting in this context?  For me, quoted text is text that is not being executed.  Thus, the Quoted class simply returns the text it received.  Therefore, I cannot understand why this class is needed.

victornoel commented:
@ruca905 no problem, I just wanted to be sure you were still around.

Usually, to quote some text means to add quotes around it, so if you have `some string`, the quoted text is `"some string"`. Is it clearer?

Maybe before trying to solve everything at once, you can try to convert only a few of the constructor calls and leave the rest in a new todo?

ruca905 commented:
@victornoel

ruca905 commented:
Hm, But it still creates a lot of duplicate code

victornoel commented:
@ruca905 let's convert a few, for example choose 3 of them, of the existing matchers to the correct constructor, and we will see from there how to remove duplication :)

victornoel commented:
@ruca905 also if you are not comfortable with the task, you can always refuse it and do another one that will be assigned to you!

ruca905 commented:
@victornoel, in my previous commits i already deleted first twice constructors in `MatcherOf` and rewrote all of matchers where they were using. Also, I introduced `Quoted` class. Is it enough to complete this PR?

victornoel commented on src/main/java/org/llorllale/cactoos/matchers/MatcherOf.java:
```diff
@@ -46,9 +46,7 @@
- * @todo #135:30min Remove all constructors except the last one so that
- *  every matcher implemented using MatcherOf take care of properly
- *  describe itself and the mismatch.
+ * @todo #204:30m/DEV Rewrite the Matcher's, which constructed from others constructor, to use the last constructor of MatcherOf. And delete all constructors except last one.
```
@ruca905 I don't think the `#204:30m/DEV` format is correct, I think it should be `#165:30min` because you are solving the issue #165, not #204!

victornoel commented on src/test/java/org/llorllale/cactoos/matchers/MatcherOfTest.java:
```diff
@@ -44,7 +47,11 @@
-            new MatcherOf<>(x -> x > 5),
+            new MatcherOf<>(x -> x > 5,
+                    desc -> desc.appendText(
+                        new UncheckedText("matches when arg satisfies the predicate").asString()
```
@ruca905 wrapping this in an `UncheckedText` is not needed, just use `appendText` with the string directly. Also don't bother about putting something complex here, just write `">5"` for example, this is a test and should stay simple since this part is not tested here

victornoel reviewed (changes requested):
@ruca905 here are some comments, also, I think you should at least update one or two existing matcher to use the correct constructor.
I'm not sure what you mean about duplicate code though…

victornoel commented on src/main/java/org/llorllale/cactoos/matchers/InputHasContent.java:
```diff
@@ -58,7 +58,9 @@ public InputHasContent(final String text) {
-                (String input) -> text.asString().equals(input), text
+                (String input) -> text.asString().equals(input),
+                    desc -> desc.appendText(text.asString()),
+                    (actual, desc) -> desc.appendValue(actual)
```
@ruca905 can you add some context in both the two functions in correcpondnce with this matcher responsibility? You can see other usages of `MatcherOf` as examples. `HasSize` is a good example :)

victornoel commented on src/main/java/org/llorllale/cactoos/matchers/EndsWith.java:
```diff
@@ -53,7 +53,8 @@ public EndsWith(final Text text) {
                     (String act) -> act.endsWith(text.asString()),
-                    text
+                        desc -> desc.appendText(text.asString()),
+                        (actual, desc) -> desc.appendValue(actual)
```
@ruca905 can you add a new todo so that we add a constructor to `TextMatcher` taking a `Func<String, Boolean>` in order to simplify this and all the other `TextMatcher`?

victornoel commented on src/main/java/org/llorllale/cactoos/matchers/MatcherOf.java:
```diff
@@ -46,9 +46,7 @@
- * @todo #135:30min Remove all constructors except the last one so that
- *  every matcher implemented using MatcherOf take care of properly
- *  describe itself and the mismatch.
+ * @todo #165:30m/DEV Rewrite the Matcher's, which constructed from others constructor, to use the last constructor of MatcherOf. And delete all constructors except last one.
```
@ruca905 please remove `/DEV` from todo specification

victornoel commented on src/main/java/org/llorllale/cactoos/matchers/Quoted.java:
```diff
@@ -0,0 +1,55 @@
+ * @todo #165:15m/DEV Move it to cactoos basic types as quoted text.
+ * This would be reused if there is the need to quote something.
+ */
+public class Quoted extends TextEnvelope {
```
@ruca905 please remove this class if you don't intend to use it

victornoel reviewed (changes requested):
@ruca905 see my comments. Sorry this is long because we went around the problem for some time, a bit more effort and we will have something already good. Also please fix the build, we won't be able to merge until then.

victornoel commented on src/test/java/org/llorllale/cactoos/matchers/EndsWithTest.java:
```diff
@@ -50,9 +50,9 @@ public void matches() {
             "The matcher gives negative result for the invalid arguments",
-            new EndsWith("!"),
+            new EndsWith("The sentence!"),
             new Mismatches<>(
-                new TextOf("The sentence."),
+                new TextOf("!"),
```
@ruca905 I believe this change is incorrect

victornoel commented on src/main/java/org/llorllale/cactoos/matchers/TextMatcher.java:
```diff
@@ -37,6 +37,15 @@
+ *             (actual, desc) -> desc.appendValue(actual)
+ *     ),
+ *     "Text ending with "
+ * )
```
@ruca905 I think you need to add one more space in front of this last parenthesis...

victornoel commented on src/main/java/org/llorllale/cactoos/matchers/MatchesRegex.java:
```diff
@@ -53,7 +53,8 @@ public MatchesRegex(final Text regex) {
                     (String act) -> act.matches(regex.asString()),
-                    regex
+                        desc -> desc.appendText(regex.asString()),
+                        (actual, desc) -> desc.appendValue(actual)
```
@ruca905 for all those matchers you modified, you need to add some relevant message to the desc in the two func here. For example here it's about a regex. Same for all the other ones (or if you don't want to do them right now, do just a few of them as I proposed earlier, but do them completely).

victornoel commented on src/main/java/org/llorllale/cactoos/matchers/MatcherOf.java:
```diff
@@ -46,9 +46,7 @@
- * @todo #135:30min Remove all constructors except the last one so that
- *  every matcher implemented using MatcherOf take care of properly
- *  describe itself and the mismatch.
+ * @todo #165:30m Rewrite the Matcher's, which constructed from others constructor, to use the last constructor of MatcherOf. And delete all constructors except last one.
```
@ruca905 I think you need to add some line return, qulice won't be happy if the line is too long

victornoel commented on src/main/java/org/llorllale/cactoos/matchers/EndsWith.java:
```diff
@@ -53,7 +53,9 @@ public EndsWith(final Text text) {
                 new MatcherOf<>(
                     (String act) -> act.endsWith(text.asString()),
-                    text
+                        desc -> desc.appendText("ends with the word " + new Quoted(text)),
```
@ruca905 please do not use the `+` but the already available methods on desc: `desc.appendText("ends with the word ").appendValue(text.asString())` which is responsible of quoting things properly.

victornoel commented on src/main/java/org/llorllale/cactoos/matchers/MatcherOf.java:
```diff
@@ -46,9 +46,8 @@
- *  every matcher implemented using MatcherOf take care of properly
- *  describe itself and the mismatch.
+ * @todo #165:30m Rewrite the Matcher's, which constructed from others constructor,
+ *  to use the last constructor of MatcherOf. And delete all constructors except last one.
```
@ruca905 I'm confused, didn't you already remove all the constructors except the last one? Please update the todo accordingly...

victornoel commented on src/main/java/org/llorllale/cactoos/matchers/TextMatcher.java:
```diff
@@ -37,6 +37,15 @@
+ *  ____________(actual, desc) -> desc.appendValue(actual)
+ *  ____),
+ *  ____"Text ending with "
+ *  )
```
@ruca905 a few things here:
- is it really needed to add those `_`?
- can you replace the example with the desired implementation instead of the current implementation
- can you also cite the name of the class this is about
- can you add something to say this should be done for all classes based on TextMatcher

victornoel commented on src/test/java/org/llorllale/cactoos/matchers/EndsWithTest.java:
```diff
@@ -50,7 +50,7 @@ public void matches() {
         new Assertion<>(
             "The matcher gives negative result for the invalid arguments",
-            new EndsWith("!"),
+            new EndsWith("The sentence!"),
```
@ruca905 can you just revert it to the original code? It was correct before I think...

victornoel commented on src/test/java/org/llorllale/cactoos/matchers/MatcherOfTest.java:
```diff
@@ -69,7 +77,11 @@ public void mismatchesFunc() {
+                    new FuncOf<>(String::trim, true),
+                    desc -> desc.appendText("Trimmed string"),
+                    (actual, desc) -> desc.appendValue(actual)
+                    ),
```
@ruca905 the indentation is incorrect in the few lines above

victornoel reviewed (changes requested):
@ruca905 sorry but there are still problems. This is why it is better to try to tackle just a little bit of the problem instead of doing everything at once :)

Also please notify me after solving the problems so that I can review as soon as you are done .

victornoel commented:
@ruca905 are you still working on tihs? Also the build is failing, it needs to pass to be merged :)

ruca905 commented:
@victornoel Yes, I'm working, but I can't figure out how mismatch tests work. Even when I try to run them for the cloned repository without my changes, they crash. As an example, this is the `mismatches()` test in `EndsWithTest`

kokodyn commented:
Hi @ruca905 
I have a similar problem with tests using class Mismaches are all failing.
Could you please paste here output from command?
mvn clean test

victornoel commented:
@kokodyn @ruca905 `Mismatches` is not fully implemented (see #138) so if the the test is correctly implemented (i.e., the failing message passed to `Mismatches` is correct), the test pass as expected, but if the test fails, the error of `Mismatches` is not very useful.

Hence you either need to wait for #138 to be fixed or you should read the code of the tested `Matcher` to know what is going to be the expected mismatch message.

ruca905 commented:
@0crat wait #138

0crat commented:
Code review was too long (98 days), architects (@victornoel) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

<!-- https://www.0crat.com/footprint/C63314D6Z/af229998-34d0-42a0-81fe-63b571a44520, version: 0.55, hash: ${buildNumber} -->

0crat commented:
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @andreoss[/z](https://www.0crat.com/u/andreoss), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

<!-- https://www.0crat.com/footprint/C63314D6Z/46598e10-aff5-4c57-a313-26994ffe0a08, version: 0.55, hash: ${buildNumber} -->

sereshqua commented:
@ruca905 please make sure you start all your comments with the name of the user they are referred to ([see](https://github.com/llorllale/cactoos-matchers/pull/197#issuecomment-700452162)) and please make sure that all problems found by code reviewers would be addressed by you in next PRs, thanks

0crat commented:
Pull request [#197](https://github.com/llorllale/cactoos-matchers/issues/197) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

<!-- https://www.0crat.com/footprint/C63314D6Z/82613fe8-d668-4658-8019-31b168b661dc, version: 0.55, hash: ${buildNumber} -->

sereshqua commented:
@0crat quality acceptable
